packages/gatsby/src/db/node-tracking.js

```diff
@@ -15,30 +15,71 @@ const getRootNodeId = node => rootNodeMap.get(node)
  * @see trackInlineObjectsInRootNode
  * @param {(Object|Array)} data Inline object or array
  * @param {string} nodeId Id of node that contains data passed in first parameter
+ * @param {boolean} sanitize Wether to strip objects of unuspported and not serializable fields
+ * @param {string} [ignore] Fieldname that doesn't need to be tracked and sanitized
+ *
  */
-const addRootNodeToInlineObject = (data, nodeId) => {
-  if (_.isPlainObject(data) || _.isArray(data)) {
-    _.each(data, o => addRootNodeToInlineObject(o, nodeId))
-    rootNodeMap.set(data, nodeId)
+const addRootNodeToInlineObject = (data, nodeId, sanitize, isNode = false) => {
+  const isPlainObject = _.isPlainObject(data)
+
+  if (isPlainObject || _.isArray(data)) {
+    let returnData = data
+    if (sanitize) {
+      returnData = isPlainObject ? {} : []
+    }
+    let anyFieldChanged = false
+    _.each(data, (o, key) => {
+      if (isNode && key === `internal`) {
+        returnData[key] = o
+        return
+      }
+      returnData[key] = addRootNodeToInlineObject(o, nodeId, sanitize)
+
+      if (returnData[key] !== o) {
+        anyFieldChanged = true
+      }
+    })
+
+    if (anyFieldChanged) {
+      if (isPlainObject) {
+        data = _.pickBy(returnData, p => p !== undefined)
+      } else {
+        data = returnData.filter(p => p !== undefined)
+      }
+    }
+
+    // don't need to track node itself
+    if (!isNode) {
+      rootNodeMap.set(data, nodeId)
+    }
+
+    // arrays and plain objects are supported - no need to to sanitize
+    return data
   }
+
+  if (sanitize && data !== null) {
```

Comment:
`typeof null === 'object'` so we don't have to do a check for !== null
```suggestion
  if (sanitize && !isTypeSupported(data) {
    return undefined;
  }
```

```js
/**
 * @param {*} data
 * @return {boolean}
 */
function isTypeSupported(data) {
  const type = typeof data
  const isSupported =
    type === `number` ||
    type === `string` ||
    type === `boolean` ||
    data instanceof Date

   return isSupported;
}
```

Reply:
addressed in https://github.com/gatsbyjs/gatsby/pull/10732/commits/843497eb3d19dca879935c6cdbe48c6eb2b38ad3 and https://github.com/gatsbyjs/gatsby/pull/10732/commits/95dc28fa7fa065b17535f78cd056abff04fec949